### DISCUSSION TITLE
How to? - Supply configuration when using MakeDefaultHandle

### DISCUSSION THREAD

**Discussion by elrob** [Ptrskay3/axum-prometheus#47] [Q&A] [answered]
I want to be able to pass some configuration which can be used in `make_default_handle` e.g. statsd host or some other settings. [README](https://github.com/ptrskay3/axum-prometheus?tab=readme-ov-file#using-a-different-exporter-than-prometheus)

Is there a way to do this? I couldn't work it out.

How can the host or the queue size or the prefix be passed in rather than hardcoded in the below example?
```rust
impl MakeDefaultHandle for Recorder {
    // We don't need to return anything meaningful from here (unlike PrometheusHandle)
    // Let's just return an empty tuple.
    type Out = ();

    fn make_default_handle() -> Self::Out {
        // The regular setup for StatsD..
        let recorder = StatsdBuilder::from("127.0.0.1", 8125)
            .with_queue_size(5000)
            .with_buffer_size(1024)
            .build(Some("prefix"))
            .expect("Could not create StatsdRecorder");

        metrics::set_boxed_recorder(Box::new(recorder)).unwrap();
    }
}
```

**Comment by Ptrskay3** [comment]
You can use [`std::sync::OnceLock`](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) (or [`OnceCell`](https://doc.rust-lang.org/std/cell/struct.OnceCell.html)) to one-time initialize these variables, then in the trait implementation you can read them. Here I'm just using an environment variable to remain simple, but you may initialize them from any source.
```rust
use std::sync::OnceLock;
static HOST: OnceLock<String> = OnceLock::new();

struct Recorder;

impl MakeDefaultHandle for Recorder {
    type Out = ();

    fn make_default_handle() -> Self::Out {
        let recorder = StatsdBuilder::from(
            HOST.get_or_init(|| std::env::var("STATSD_HOST").expect("to be set")),
            8125,
        )
        .with_queue_size(5000)
        .with_buffer_size(1024)
        .build(Some("prefix"))
        .expect("Could not create StatsdRecorder");

        metrics::set_global_recorder(recorder).unwrap();
    }
}
```

EDIT: Honestly, to me this part seems a bit awkward or convoluted to use, since the expected solution would be just adding fields to the `Recorder` struct, and then referencing it in the trait. But the `GenericMetricLayer::pair` API doesn't really give options to initialize the struct manually..  Maybe it's worth to explore this space, and perhaps something productive comes out of this. I'll convert this to a discussion, and create a new API design issue.

**Reply by elrob** [reply]
Thank you! Yes, I realized it would be possible with an env var like that. However, I'd like to be able to pass in some configuration, as the location where environment values and configuration files are read is separate and I'd like to keep it that way.

**Reply by Ptrskay3** [reply]
I completely forgot to mention that you can use the builder API like so:
```rust
struct Recorder;

impl MakeDefaultHandle for Recorder {
    type Out = ();

    fn make_default_handle() -> Self::Out {
        // The regular setup for StatsD..
        let recorder = StatsdBuilder::from("127.0.0.1", 8125)
            .with_queue_size(5000)
            .with_buffer_size(1024)
            .build(Some("prefix"))
            .expect("Could not create StatsdRecorder");

        metrics::set_global_recorder(recorder).unwrap();
    }
}

#[tokio::main]
async fn main() {
    let host = "127.0.0.1";

    let (metric_layer, _) = axum_prometheus::MetricLayerBuilder::<'_, _, Recorder, _>::new()
        .with_metrics_from_fn(move || {
            let recorder = StatsdBuilder::from(host, 8125)
                .with_queue_size(5000)
                .with_buffer_size(1024)
                .build(Some("prefix"))
                .expect("Could not create StatsdRecorder");

            metrics::set_global_recorder(recorder).unwrap();
        })
        .build_pair();
    // ...
}
```

The downside of this approach is that you still need to provide the `impl MakeDefaultHandle for Recorder`, which is completely redundant.  The conclusion remains the same, this API surely needs improvement.

**Comment by Ptrskay3** [comment, marked as answer]
This is now possible, because `make_default_handle` takes self as argument, and also there's `GenericMetricLayer::pair_from`.
